File: pkg/contract/primitive.go
// SPDX-License-Identifier: AGPL-3.0-only

package contract

import (
	"context"
	"time"
)

// PrimitiveLayer is the execution-layer boundary. It returns structured
// observations and file/content references, never raw frame buffers.
type PrimitiveLayer interface {
	ConnectDevice(context.Context, DeviceConnectRequest) (DeviceSession, error)
	StartApp(context.Context, AppRequest) (ActionResult, error)
	StopApp(context.Context, AppRequest) (ActionResult, error)
	Capture(context.Context, CaptureRequest) (CaptureRef, error)
	Match(context.Context, MatchRequest) (MatchResult, error)
	OCR(context.Context, OCRRequest) (OCRResult, error)
	GetColor(context.Context, ColorRequest) (ColorResult, error)
	Tap(context.Context, TapRequest) (ActionResult, error)
	Swipe(context.Context, SwipeRequest) (ActionResult, error)
	WaitFor(context.Context, WaitForRequest) (WaitForResult, error)
}

type DeviceConnectRequest struct {
	ProfileID ProfileID         `json:"profileId"`
	DeviceID  string            `json:"deviceId"`
	Backend   string            `json:"backend"`
	Metadata  map[string]string `json:"metadata,omitempty"`
	Timeout   time.Duration     `json:"timeout"`
}

type DeviceSession struct {
	ID          string            `json:"id"`
	DeviceID    string            `json:"deviceId"`
	Backend     string            `json:"backend"`
	Resolution  Resolution        `json:"resolution"`
	ConnectedAt time.Time         `json:"connectedAt"`
	Metadata    map[string]string `json:"metadata,omitempty"`
}

type AppRequest struct {
	SessionID string        `json:"sessionId"`
	Package   string        `json:"package"`
	Activity  string        `json:"activity,omitempty"`
	Timeout   time.Duration `json:"timeout"`
}

type CaptureRequest struct {
	SessionID string        `json:"sessionId"`
	Region    *Rect         `json:"region,omitempty"`
	Timeout   time.Duration `json:"timeout"`
	Reason    string        `json:"reason"`
}

type CaptureRef struct {
	ID         string     `json:"id"`
	ImageRef   string     `json:"imageRef"`
	ImageHash  string     `json:"imageHash,omitempty"`
	Resolution Resolution `json:"resolution"`
	CapturedAt time.Time  `json:"capturedAt"`
}

type MatchRequest struct {
	SessionID  string        `json:"sessionId"`
	CaptureID  string        `json:"captureId,omitempty"`
	Templates  []TemplateRef `json:"templates"`
	Region     *Rect         `json:"region,omitempty"`
	Threshold  float64       `json:"threshold"`
	MaxResults int           `json:"maxResults,omitempty"`
	Timeout    time.Duration `json:"timeout"`
}

type TemplateRef struct {
	ID         string     `json:"id"`
	Path       string     `json:"path"`
	Hash       string     `json:"hash,omitempty"`
	Game       GameKey    `json:"game"`
	Server     ServerKey  `json:"server"`
	Locale     string     `json:"locale,omitempty"`
	Resolution Resolution `json:"resolution"`
}

type MatchResult struct {
	Hits       []MatchHit `json:"hits"`
	ObservedAt time.Time  `json:"observedAt"`
}

type MatchHit struct {
	TemplateID string  `json:"templateId"`
	Score      float64 `json:"score"`
	Rect       Rect    `json:"rect"`
}

type OCRRequest struct {
	SessionID string        `json:"sessionId"`
	CaptureID string        `json:"captureId,omitempty"`
	Region    Rect          `json:"region"`
	Languages []string      `json:"languages,omitempty"`
	Timeout   time.Duration `json:"timeout"`
}

type OCRResult struct {
	Text       string         `json:"text"`
	Blocks     []OCRBlock     `json:"blocks,omitempty"`
	Confidence float64        `json:"confidence,omitempty"`
	ObservedAt time.Time      `json:"observedAt"`
	Warnings   []RuntimeError `json:"warnings,omitempty"`
}

type OCRBlock struct {
	Text       string  `json:"text"`
	Rect       Rect    `json:"rect"`
	Confidence float64 `json:"confidence,omitempty"`
}

type ColorRequest struct {
	SessionID string        `json:"sessionId"`
	CaptureID string        `json:"captureId,omitempty"`
	Point     Point         `json:"point"`
	Timeout   time.Duration `json:"timeout"`
}

type ColorResult struct {
	RGBA       string    `json:"rgba"`
	ObservedAt time.Time `json:"observedAt"`
}

type TapRequest struct {
	SessionID string        `json:"sessionId"`
	Point     Point         `json:"point"`
	Reason    string        `json:"reason"`
	Timeout   time.Duration `json:"timeout"`
}

type SwipeRequest struct {
	SessionID string        `json:"sessionId"`
	From      Point         `json:"from"`
	To        Point         `json:"to"`
	Duration  time.Duration `json:"duration"`
	Reason    string        `json:"reason"`
	Timeout   time.Duration `json:"timeout"`
}

type WaitForRequest struct {
	SessionID string        `json:"sessionId"`
	Condition string        `json:"condition"`
	Timeout   time.Duration `json:"timeout"`
	PollEvery time.Duration `json:"pollEvery"`
}

type WaitForResult struct {
	Satisfied  bool      `json:"satisfied"`
	ObservedAt time.Time `json:"observedAt"`
	Details    string    `json:"details,omitempty"`
}

type ActionResult struct {
	OK         bool              `json:"ok"`
	ObservedAt time.Time         `json:"observedAt"`
	Error      *RuntimeError     `json:"error,omitempty"`
	Metadata   map[string]string `json:"metadata,omitempty"`
}

type Point struct {
	X int `json:"x"`
	Y int `json:"y"`
}

type Rect struct {
	X      int `json:"x"`
	Y      int `json:"y"`
	Width  int `json:"width"`
	Height int `json:"height"`
}

File: contracts/rust/actingcommand-contract-rs/src/primitive.rs
// SPDX-License-Identifier: AGPL-3.0-only

//! Rust backup boundary for low-level device, capture, recognition, and input primitives.

use crate::types::*;

/// Execution-layer boundary mirrored from Go `PrimitiveLayer`.
pub trait PrimitiveLayer {
    fn connect_device(
        &mut self,
        ctx: &RuntimeContext,
        request: DeviceConnectRequest,
    ) -> ContractResult<DeviceSession>;

    fn start_app(
        &mut self,
        ctx: &RuntimeContext,
        request: AppRequest,
    ) -> ContractResult<ActionResult>;

    fn stop_app(
        &mut self,
        ctx: &RuntimeContext,
        request: AppRequest,
    ) -> ContractResult<ActionResult>;

    fn capture(
        &mut self,
        ctx: &RuntimeContext,
        request: CaptureRequest,
    ) -> ContractResult<CaptureRef>;

    fn match_templates(
        &mut self,
        ctx: &RuntimeContext,
        request: MatchRequest,
    ) -> ContractResult<MatchResult>;

    fn ocr(&mut self, ctx: &RuntimeContext, request: OcrRequest) -> ContractResult<OcrResult>;

    fn get_color(
        &mut self,
        ctx: &RuntimeContext,
        request: ColorRequest,
    ) -> ContractResult<ColorResult>;

    fn tap(&mut self, ctx: &RuntimeContext, request: TapRequest) -> ContractResult<ActionResult>;

    fn swipe(
        &mut self,
        ctx: &RuntimeContext,
        request: SwipeRequest,
    ) -> ContractResult<ActionResult>;

    fn wait_for(
        &mut self,
        ctx: &RuntimeContext,
        request: WaitForRequest,
    ) -> ContractResult<WaitForResult>;
}

#[derive(Debug, Clone, PartialEq)]
pub struct DeviceConnectRequest {
    pub profile_id: ProfileId,
    pub device_id: String,
    pub backend: String,
    pub metadata: Metadata,
    pub timeout_ms: DurationMillis,
}

#[derive(Debug, Clone, PartialEq)]
pub struct DeviceSession {
    pub id: String,
    pub device_id: String,
    pub backend: String,
    pub resolution: Resolution,
    pub connected_at: Timestamp,
    pub metadata: Metadata,
}

#[derive(Debug, Clone, PartialEq)]
pub struct AppRequest {
    pub session_id: String,
    pub package: String,
    pub activity: Option<String>,
    pub timeout_ms: DurationMillis,
}

#[derive(Debug, Clone, PartialEq)]
pub struct CaptureRequest {
    pub session_id: String,
    pub region: Option<Rect>,
    pub timeout_ms: DurationMillis,
    pub reason: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct CaptureRef {
    pub id: String,
    pub image_ref: String,
    pub image_hash: Option<String>,
    pub resolution: Resolution,
    pub captured_at: Timestamp,
}

#[derive(Debug, Clone, PartialEq)]
pub struct MatchRequest {
    pub session_id: String,
    pub capture_id: Option<String>,
    pub templates: Vec<TemplateRef>,
    pub region: Option<Rect>,
    pub threshold: f64,
    pub max_results: Option<i32>,
    pub timeout_ms: DurationMillis,
}

#[derive(Debug, Clone, PartialEq)]
pub struct TemplateRef {
    pub id: String,
    pub path: String,
    pub hash: Option<String>,
    pub game: GameKey,
    pub server: ServerKey,
    pub locale: Option<String>,
    pub resolution: Resolution,
}

#[derive(Debug, Clone, PartialEq)]
pub struct MatchResult {
    pub hits: Vec<MatchHit>,
    pub observed_at: Timestamp,
}

#[derive(Debug, Clone, PartialEq)]
pub struct MatchHit {
    pub template_id: String,
    pub score: f64,
    pub rect: Rect,
}

#[derive(Debug, Clone, PartialEq)]
pub struct OcrRequest {
    pub session_id: String,
    pub capture_id: Option<String>,
    pub region: Rect,
    pub languages: Vec<String>,
    pub timeout_ms: DurationMillis,
}

#[derive(Debug, Clone, PartialEq)]
pub struct OcrResult {
    pub text: String,
    pub blocks: Vec<OcrBlock>,
    pub confidence: Option<f64>,
    pub observed_at: Timestamp,
    pub warnings: Vec<RuntimeError>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct OcrBlock {
    pub text: String,
    pub rect: Rect,
    pub confidence: Option<f64>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ColorRequest {
    pub session_id: String,
    pub capture_id: Option<String>,
    pub point: Point,
    pub timeout_ms: DurationMillis,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ColorResult {
    pub rgba: String,
    pub observed_at: Timestamp,
}

#[derive(Debug, Clone, PartialEq)]
pub struct TapRequest {
    pub session_id: String,
    pub point: Point,
    pub reason: String,
    pub timeout_ms: DurationMillis,
}

#[derive(Debug, Clone, PartialEq)]
pub struct SwipeRequest {
    pub session_id: String,
    pub from: Point,
    pub to: Point,
    pub duration_ms: DurationMillis,
    pub reason: String,
    pub timeout_ms: DurationMillis,
}

#[derive(Debug, Clone, PartialEq)]
pub struct WaitForRequest {
    pub session_id: String,
    pub condition: String,
    pub timeout_ms: DurationMillis,
    pub poll_every_ms: DurationMillis,
}

#[derive(Debug, Clone, PartialEq)]
pub struct WaitForResult {
    pub satisfied: bool,
    pub observed_at: Timestamp,
    pub details: Option<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ActionResult {
    pub ok: bool,
    pub observed_at: Timestamp,
    pub error: Option<RuntimeError>,
    pub metadata: Metadata,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Point {
    pub x: i32,
    pub y: i32,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Rect {
    pub x: i32,
    pub y: i32,
    pub width: i32,
    pub height: i32,
}
